### DISCUSSION TITLE
Streaming tool-use message parser?

### DISCUSSION THREAD

**Discussion by adamavenir** [modelcontextprotocol/.github#110] [General]
### Pre-submission Checklist

- [X] I have verified that this discussion would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Discussion Topic

Assuming one doesn't already exist (I can't find one), I'm building a streaming parser for Claude messages which is probably going to be useful for others. 

Here's my basic goal: 

- Process Claude's responses token by token in real-time, returning individual messages separately formatted as appropriate
- Automatically detects and separates different message components
When Claude uses tools, messages are automatically split into:
- Regular messages
- Thought process messages
- Tool execution messages (with parameters)

Example Flow:
```
Claude: Let me examine that repository.
---
Claude: 💭 *Let me search for this repository to get basic information about it*
---
Claude: 🛠️ *using `search_repositories` tool*
       repository: `adamavenir/reponame`
---
Claude: 🛠️ *using `read_multiple_files` tool*
       repository: `adamavenir/reponame`
       files: [`package.json`, `README.md`, `.gitignore`, `server.js`]
---
Claude: Based on my analysis of the repository, here are the key findings...
```

Looking for feedback and input... and if someone already built a standalone module for this, I'll just use it!

**Comment by jspahrsummers** [comment]
I'm afraid this doesn't seem related to MCP specifically, but rather to building an Anthropic API client (though please correct me if that's wrong). As such, I'm going to close this out.

**Reply by adamavenir** [reply]
It seems highly related to anyone building an MCP client, but I'm totally fine with closing it.
